### PR TITLE
OpenJ9 Adopts RI java.lang.String

### PR DESCRIPTION
**Pull request by JasonFengJ9**
Add OpenJ9 specific fields/methods into RI `java.lang.String`;
Make `StringCoding.implEncodeISOArray()` `public` which is required by RI `String`;
A separated PR (https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/296) is to remove legacy `StringCoding` code.

Note: this change to `java.lang.String` has no impact on current JDK17 build cause it is not included by OpenJ9 yet.

This unlocks https://github.com/eclipse/openj9/pull/12209.

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

**Comment by JasonFengJ9** [comment]
`11:25:54  ERROR - should have LF line terminators: 'src/utils/IdealGraphVisualizer/ServerCompiler/nbproject/project.xml' type: 'XML 1.0 document, ASCII text, with CRLF line terminators'`

jenkins line endings check

**Comment by JasonFengJ9** [comment]
jenkins copyright check

**Comment by pshipton** [comment]
@fjeremic any concerns with this?

**Comment by pshipton** [comment]
jenkins compile osx jdknext

**Comment by JasonFengJ9** [comment]
> @fjeremic any concerns with this?

Somehow I can't request Filip as a reviewer. (DM him)

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -217,11 +224,389 @@
...
+
+			// Copy in s1 contents
+			decompressedArrayCopy(s1.value, 0, value, 0, s1len);
+
+			if (COMPACT_STRINGS) {
+				initCompressionFlag();
+			}
+		}
+	}
+
```
Comment (fjeremic) [inline src/java.base/share/classes/java/lang/String.java:595]: I think we're missing a private constructor accelerated by the JIT:
https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/jcl/src/java.base/share/classes/java/lang/String.java#L1156-L1337

Comment (JasonFengJ9) [reply src/java.base/share/classes/java/lang/String.java:595]: Good catch, just wondering if JIT falls back to a less-optimized call or we were just lucky not hit by `NoSuchMethodError`?

Comment (fjeremic) [reply src/java.base/share/classes/java/lang/String.java:595]: We have an optimization which looks for String concatenations and replaces them with a call to a private String constructor. If that constructor is not found we just don't do the optimization (StringPeepholes).

Comment (JasonFengJ9) [reply src/java.base/share/classes/java/lang/String.java:595]: Filip, could you shine some lights on how the fallback works when this constructor [1] is not found? A quick experiment shows that missing a similar 3 parameter private constructor [2] caused build failure
```
Assertion failed at /Users/ivm/osx-openj9-builds/jdk17-0316-jdk17ristrV2/openj9-openjdk-jdk/omr/compiler/compile/OMRSymbolReferenceTable.cpp:979: method
VMState: 0x000535ff
	methodSymRefFromName: method must exist: java/lang/String.<init>(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V
compiling java/util/Arrays.copyOfRange([BII)[B at level: warm
``` 

[1] https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/jcl/src/java.base/share/classes/java/lang/String.java#L1156-L1159
[2] https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/jcl/src/java.base/share/classes/java/lang/String.java#L1007-L1010

Comment (fjeremic) [reply src/java.base/share/classes/java/lang/String.java:595]: This looks like the code which tries to find the symbol reference:
https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/runtime/compiler/optimizer/StringPeepholes.cpp#L157-L214

I did not know that call to `methodSymRefFromName` will assert if it cannot find the symbol. So it looks like it is a bug not to have these methods.

Comment (JasonFengJ9) [reply src/java.base/share/classes/java/lang/String.java:682]: Added `private String(int v1, String s1, int v2, String s2, String s3)` for https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/runtime/compiler/optimizer/StringPeepholes.cpp#L198 
Also added `String(byte[] data, int start, int length, boolean compressed)` for https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/runtime/compiler/optimizer/StringPeepholes.cpp#L197

@fjeremic please have another look.

**Comment by JasonFengJ9** [comment]
jenkins compile zlinux jdknext

**Comment by fjeremic** [comment]
Functionally this is ok. Performance wise it is questionable since it is missing things like hashcode acceleration, etc. We should discuss this on Monday on the call you scheduled before proceeding.

**Comment by JasonFengJ9** [comment]
Rebasing to resolve the conflict.
It appears the modification to `equals()` & `hashCode()` has more chances of conflict.

**Comment by JasonFengJ9** [comment]
@fjeremic could you take another look at this? thanks.

**Review by fjeremic — APPROVED** [review]

**Comment by JasonFengJ9** [comment]
@pshipton could you review & merge if approve?

**Comment by JasonFengJ9** [comment]
jenkins compile zlinux jdknext

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdknext

File: src/java.base/share/classes/java/lang/StringCoding.java [hunk 2]
```diff
@@ -600,7 +606,7 @@ private static Result decodeLatin1(byte[] ba, int off, int len) {
     }
 
     @IntrinsicCandidate
-    private static int implEncodeISOArray(byte[] sa, int sp,
+    public static int implEncodeISOArray(byte[] sa, int sp,
```
Comment (pshipton) [inline src/java.base/share/classes/java/lang/StringCoding.java:609]: Is this change needed?

Comment (JasonFengJ9) [reply src/java.base/share/classes/java/lang/StringCoding.java:609]: Yes, this is to match `StringCoding` in `master` branch to be referred by `String`.

**Review by pshipton — APPROVED** [review]
